### ISSUE TITLE
No information if SubscriberServiceApiClient.AcknowledgeAsync has exceeded deadline

### ISSUE THREAD

**Issue by Rassi** [googleapis/google-cloud-dotnet#7425]
TLDR: No exception or return value to signal if acknowledgement deadline has been exceeded, calling SubscriberServiceApiClient.AcknowledgeAsync, resulting in many retries, or deadletter messages, and no way of identifying a potential problem in message handling.

#### Issue
We are switching from a different message queue system to PubSub, and are using SubscriberServiceApiClient to ensure similar operation with minimal codechanges. Our previous message queue API had simple "pull/delete" methods, and our infrastructure is built up around many instances running at the same time (Kubernetes). This allowed each system instance to handle one message at a time. This is the reason why we've implemented SubscriberServiceApiClient since we can control only processing one message at a time.

We are now experiencing one of our subscriptions growing over the last couple of days, and have identified that message handling has been experiencing problems which prolonged it to the point where the acknowledgement deadline has been exceeded. Our old message queue would throw an exception when trying to delete/acknowledge a message which had exceeded it's timeout/deadline, but this does not seem to be the case with PubSub.

How do we identify if we exceed the Acknowledgement Deadline which has been set on the subscription?

#### Steps to reproduce

  1. Create subscription with an acknowledgement deadline of 10 seconds.
  2. Reserve message using SubscriberServiceApiClient.PullAsync.
  3. Wait more than 10 seconds.
  4. Observe that SubscriberServiceApiClient.AcknowledgeAsync does not throw exception or return response signaling acknowledgement deadline exceeded.

**Comment by jskeet** [comment]
Thanks for raising the issue. I *suspect* this is a question more about the API than about the .NET client library, so I'll ask internally and get back to you when I have an answer.

**Comment by kamalaboulhosn** [comment]
The ack deadline is not a strong guarantee either way: there is no guarantee that a message won't be delivered before the ack deadline has expired and there is no guarantee that an ack that comes after the deadline won't be accepted. In general, acks are meant to be fire-and-forget operations for the client and in the server, they are not processed synchronously when received. At this time, there is no way to get a response from an ack and know definitively if it was accepted or not. This is part of Pub/Sub's at-least-once delivery model.

**Comment by Rassi** [comment]
Thank you for your swift replies, although it was not what I hoped. It is important that we are able to raise an alert if message processing takes longer than expected. Especially when it results in messages being sent back to the queue and retried, possibly resulting in processing being backed up.

Being able to use a message queue's built in deadline/timeout system has served us very well in the past. We will have to see if there is another way to identify it with PubSub.

**Comment by jskeet** [comment]
Thanks for the feedback. We'll make sure the product team is aware of it - that's never a guarantee of a change of course, but the more we know about customer requirements, the better we're able to meet them in the future.

I'll close this issue now as it's not something that can be addressed within the .NET client library.
